ArduPlane: tailsitter thrust boost proportional to pitch/yaw error

kd0aij opened the pull request:
implemented in tailsitter.cpp in order to have access to attitude target values

Observation: there must be sufficient thrust margin available to boost pitch/yaw thrust with application of throttle boost. The Stryker becomes marginal on pitch authority if, for example, one uses a weak LiPo battery. Servo performance is also important, the original servos were probably overstressed since they both failed at the same time. 

The following plot shows that throttle boost is coming in each time the elevons saturate, and that this is effective in reducing tracking error:

![thrcomp4](https://cloud.githubusercontent.com/assets/2300221/26726910/0a5c6916-4762-11e7-8d4d-f78122c334bc.png)

log file: https://drive.google.com/open?id=0Bw3digSMQXDuZGJ3aE1GekZMQWc

kd0aij commented:
Changing from 3S/2200 to 4S/1550 battery allows hover at 60% throttle (17.5A/250W), providing more margin for thrust boost. But this resulted in pitch oscillations on recovery from a large deviation. Adding gain scheduling eliminated the oscillations. The elevons are still saturating, but with thrust boost it eventually corrects the pitch error. There seems to be a lot of latency in the pitch error calculation; performance could be significantly improved if that can be speeded up.

![thrust_boost_gain_sched_4s_107](https://user-images.githubusercontent.com/2300221/27063869-7faf2378-4fb1-11e7-9a2b-be0ec969743a.png)

tridge commented:
I just did a quick test of this on the AddictionX in FlightAxis and it worked nicely, well done!
There are a couple of structural issues that I'll add individual comments on

tridge commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -774,13 +774,9 @@ void Plane::update_flight_mode(void)
         nav_roll_cd = constrain_int32(nav_roll_cd, -roll_limit, roll_limit);
         float pitch_input = channel_pitch->norm_input();
-        // Scale from normalized input [-1,1] to centidegrees
-        if (quadplane.tailsitter_active()) {
```
any reason to remove these comments?

kd0aij replied:
this looks like a merge failure; will fix

tridge commented on ArduPlane/tailsitter.cpp:
```diff
@@ -62,6 +62,61 @@ void QuadPlane::tailsitter_output(void)
     plane.pitchController.reset_I();
     plane.rollController.reset_I();
 
+    static float orig_kP = -1.0f;
```
I'd rather have this variable inside the tailsitter structure

kd0aij replied:
just keeping the gain scheduling hack localized here for now

tridge commented on ArduPlane/tailsitter.cpp:
```diff
@@ -62,6 +62,61 @@ void QuadPlane::tailsitter_output(void)
+
+        float elevon_L = SRV_Channels::get_output_norm(SRV_Channel::k_elevon_left);
+        float elevon_R = SRV_Channels::get_output_norm(SRV_Channel::k_elevon_right);
+
```
getting the elevon outputs won't work on all tailsitters - for example it doesn't work properly on a 3D plane. Instead you ask for the k_aileron and k_elevator and check for saturation on those then it will work for all aircraft, as elevons are derived from those.

kd0aij replied:
interesting; since aileron and elevator demand are summed into elevon ouputs, an elevon may saturate when neither aileron nor elevator is saturated.

tridge commented on ArduPlane/tailsitter.cpp:
```diff
@@ -62,6 +62,61 @@ void QuadPlane::tailsitter_output(void)
+
+        // reduce pitch PID gains in proportion to throttle boost
+        float pitch_kP = orig_kP / thr_boost;
+        attitude_control->get_rate_pitch_pid().kP(pitch_kP);
```
I'd like to get feedback from Leonard on if its OK to change these gains in this way

kd0aij replied:
@lthall Is this approach good enough for flight testing, and what would be the proper way to do gain scheduling?

kd0aij commented:
Yaw control was terrible with elevon_margin = 1 - max(elevon_L, elevon_R)
Changed this to 1 - .5 max(elevon_L, elevon_R) and yaw control looks better, though it may now be necessary to do gain scheduling on yaw rate kP also:
![tcomp_gsched_110](https://user-images.githubusercontent.com/2300221/27803808-a844332c-5fe8-11e7-8aa4-42ebaa92443c.png)

kd0aij commented:
Changed the hover-mode yaw (aileron) attenuation to depend on k_elevator demand as recommended by @tridge and manual control in QSTABILIZE is looking very good in calm conditions.
![tcomp_gsched_114](https://user-images.githubusercontent.com/2300221/27844713-79b8c568-60e2-11e7-8c60-8289709fe494.png)

Rebased on master and force-pushed

kd0aij commented:
Roughly real-time Google Earth animation of log 114 above:
https://drive.google.com/open?id=0Bw3digSMQXDuSVN3QTBqakZMOEk
Simulated wind by flying forward and backward during last half of test; attitude tracking including earth-frame yaw was good at pitch angles up to +/-25 degrees.

kd0aij commented:
@tridge The last commit adds throttle boost to correct a problem which showed up with the simulated AddictionX (FlightAxis). Under certain conditions there is insufficient rudder authority as demonstrated in this RF7.5 recording running recent master: https://drive.google.com/open?id=0Bw3digSMQXDueFpjdVJlbFZmTXc

This branch corrects the problem by boosting the throttle (when roll error is large) in a fashion similar to that used for the dual-motor (non-vectored) tailsitter: https://drive.google.com/open?id=0Bw3digSMQXDuVy1ITFA0V1gwOWs

I'm using the RF7.5 recording capability here since it has no significant impact on frame rate, maintaining simulation fidelity. The "gadgets" slow the sim a bit, but FPS is still close to 200. I haven't yet tried this with a real Addiction, but should soon be able to test it on a Multiplex Acromaster which should have similar flight characteristics.

kd0aij commented:
The parameter values I'm using for FlightAxis are here: https://drive.google.com/open?id=0Bw3digSMQXDubU5YSTR5VnBrVmM

tridge commented on ArduPlane/tailsitter.cpp:
```diff
@@ -62,6 +62,139 @@ void QuadPlane::tailsitter_output(void)
+    static float orig_pkP = -1.0f;
+    static float orig_rkP = -1.0f;
+    static float orig_ykP = -1.0f;
+
```
I'd like the static variables to be moved into a tcomp structure in the QuadPlane object. We don't instantiate that object at all for normal fixed wing, and it is nice to not then use the memory.

tridge commented on ArduPlane/tailsitter.cpp:
```diff
@@ -62,6 +62,139 @@ void QuadPlane::tailsitter_output(void)
+    static float orig_ykP = -1.0f;
+
+    if (tailsitter.tcomp_gain > 0) {
+        if (orig_pkP < 0.0f) {
```
I think the big lump of code here could be moved to a new function, maybe tailsitter_tcomp() ?

tridge commented on libraries/AP_Motors/AP_MotorsTailsitter.cpp:
```diff
@@ -101,21 +101,61 @@ void AP_MotorsTailsitter::output_to_motors()
+//    // override boost at low throttle demand (for safety)
+//    if (get_throttle() < (0.5f * _throttle_hover)) {
+//        throttle_thrust *= (get_throttle() / (0.5f * _throttle_hover));
+//    }
```
could we remove the commented out code?
